lib/ttfunk/table/os2.rb

```diff
@@ -40,36 +41,257 @@ class OS2 < Table
+              os2.y_superscript_y_offset, os2.y_strikeout_size,
+              os2.y_strikeout_position, os2.family_class
+            ].pack('n*')
+
+            result << os2.panose
+
+            new_char_range = unicode_blocks_for(os2, os2.char_range, subset)
+            result << BinUtils
+                      .slice_int(
+                        new_char_range.value,
+                        bit_width: 32,
+                        slice_count: 4
+                      )
+                      .pack('N*')
+
+            result << os2.vendor_id
+
+            new_cmap_table = subset.new_cmap_table[:charmap]
+            code_points = new_cmap_table
+                          .keys
+                          .select { |k| new_cmap_table[k][:new] > 0 }
+                          .sort
+
+            # "This value depends on which character sets the font supports.
+            # This field cannot represent supplementary character values
+            # (codepoints greater than 0xFFFF). Fonts that support
+            # supplementary characters should set the value in this field
+            # to 0xFFFF."
+            first_char_index = [code_points.first || 0, UNICODE_MAX].min
+            last_char_index = [code_points.last || 0, UNICODE_MAX].min
+
+            result << [
+              os2.selection, first_char_index, last_char_index
+            ].pack('n*')
+
+            if os2.version > 0
+              result << [
+                os2.ascent, os2.descent, os2.line_gap,
+                os2.win_ascent, os2.win_descent
+              ].pack('n*')
+
+              result << BinUtils
+                        .slice_int(
+                          code_pages_for(subset).value,
+                          bit_width: 32,
+                          slice_count: 2
+                        )
+                        .pack('N*')
+
+              if os2.version > 1
+                result << [
+                  os2.x_height, os2.cap_height, os2.default_char,
+                  os2.break_char, os2.max_context
+                ].pack('n*')
+              end
+            end
+          end
+        end
+
+        private
+
+        def code_pages_for(subset)
+          field = BitField.new(0)
+          return field if subset.unicode?
+
+          field.on(CODE_PAGE_BITS[subset.code_page])
+          field
+        end
+
+        def unicode_blocks_for(os2, original_field, subset)
+          field = BitField.new(0)
+          return field unless subset.unicode?
+          subset_code_points = Set.new(subset.new_cmap_table[:charmap].keys)
+          original_code_point_groups = group_original_code_points_by_bit(os2)
+
+          original_code_point_groups.each do |bit, code_points|
+            next if original_field.off?(bit)
+
+            # only flip the bit on if the subset includes all the characters
+            # that were present for this block in the original font
+            if code_points.all? { |cp| subset_code_points.include?(cp) }
```

Comment:
Yes, I see what you're saying. Before submitting the PR I tried a different approach where bits were flipped on if _any_ of the corresponding characters were present in the subset. I abandoned it because it was fairly slow and didn't seem correct (after all, the font's definition of a usable unicode range won't have been satisfied). I don't believe the font will be broken if no unicode ranges are indicated, so only flipping bits on if all characters are present seemed like the best approach. If you feel strongly about going back to the original technique, I'd be happy to revert.

Reply:
> it was fairly slow

I for some reason doubt that `any?` is slower than `all?`. They're the same in the worst case scenario.

> the font's definition of a usable unicode range won't have been satisfied

It may be not for a regular font since the font author has no idea how the font would be used. In general case the flag should be set when a code page has a decent if not full coverage of a code page.

Subset fonts are used in a very specific scenarios. In case of Prawn (and PDF in general) the fonts are only usable with a specific text. There's no risk of encountering a situation where subset font won't cover what original could have. So it still satisfies "usability" criteria to the same extent the original font would have.

I think the appropriate approach would be to unset bits for ranges that are completely removed from the font.

Reply:
> I for some reason doubt that any? is slower than all?. They're the same in the worst case scenario.

No, that's not what I meant. My original algorithm used `Array#bsearch` for every character.

> In general case the flag should be set when a code page has a decent if not full coverage of a code page.

Ok, that makes sense to me.

Reply:
Looks like this is as simple as using `any?` instead of `all?`. Cool 😎